pkg/ring/lifecycler.go

```diff
@@ -673,6 +967,21 @@ func (i *Lifecycler) processShutdown(ctx context.Context) {
 	time.Sleep(i.cfg.FinalSleep)
 }
 
+// getDenormalisedRing is a helper method to grab the ring, denormalise the
+// tokens and sort them.
+func (i *Lifecycler) getDenormalisedRing(ctx context.Context) (*Desc, error) {
+	d, err := i.KVStore.Get(ctx, ConsulKey)
+	if err != nil {
+		return nil, err
+	}
+	desc, ok := d.(*Desc)
+	if !ok {
+		return nil, fmt.Errorf("could not convert ring to Desc")
+	}
+	desc.Tokens = migrateRing(desc)
+	return desc, nil
+}
```

Comment:
`updateConsul` would have already just been called any time this function is invoked. We can save the most recently `CAS`'d value and use that instead here. The stored value can be denormalized so the TokenChecker can also use it.